trpcultivate_phenotypes/src/Service/TripalCultivatePhenotypesFileTemplateService.php

```diff
@@ -1,94 +1,137 @@
-    // MIME: TSV type file.
-    $filemime = 'text/tab-separated-values';
-
-    // Personalize the filename by appending display name of the current user, but first
-    // sanitize it by replacing all spaces into a dash character.
+    $file_extension = $file_extensions[0];
+    // See referenced file in the doc block about the mapping variables used.
+    // File MIME type.
+    // @todo Should move the following 2 methods to a more generic class
+    // instead of keeping them in FileTypes and ValidatorBase which are meant
+    // for validators.
+    $file_mime_type = self::$extension_to_mime_mapping[$file_extension];
+    // File delimiter.
+    $file_delimiter = TripalCultivatePhenotypesValidatorBase::$mime_to_delimiter_mapping[$file_mime_type[0]];
+
+    // Personalize the filename by appending display name of the current user,
+    // but first sanitize it by replacing all spaces into a dash character.
     $display_name = $this->user->getDisplayName() ?? 'anonymous-user';
     $user_display_name = str_replace(' ', '-', $display_name);
 
     // Filename: importer id - data collection template file - username . (TSV).
-    $filename = $importer_id . '-data-collection-template-file-' . $user_display_name . '.' . $fileextension;
+    $filename = $importer_id . '-data-collection-template-file-' . $user_display_name . '.' . $file_extension;
 
     // Create the file.
     $file = File::create([
       'filename' => $filename,
-      'filemime' => $filemime,
-      'uri' => $dir_template_file . $filename
+      'filemime' => $file_mime_type,
+      'uri' => $dir_template_file . $filename,
     ]);
 
     // Mark file for deletion during a Drupal maintenance.
     $file->set('status', 0);
-    // Save.
-    $file->save();
 
     // Write the contents: headers into the file created and serve the path back
-    // to the calling Importer as value to the href attribute of link to download a template file.
-
-    // File uri of the created file.
+    // to the calling Importer as value to the href attribute of link to
+    // download a template file. File uri of the created file.
     $fileuri = $file->getFileUri();
 
-    // Before we can write contents, we need to ensure the upper level folders exist.
+    // Before we can write contents, we need to ensure the upper level
+    // folders exist.
     if (!file_exists($dir_template_file)) {
-      mkdir($dir_template_file, 0777, true);
+      mkdir($dir_template_file, 0777, TRUE);
     }
 
-    // Convert the headers array into a tsv string value and post into the first line of the file.
-    $fileheaders = implode("\t", $column_headers) . "\n# DELETE THIS LINE --- START DATA HERE AND USE TAB KEY #";
+    // Convert the headers array into a delimited string value and post into the
+    // first line of the file.
+    $fileheaders = implode($file_delimiter[0], $column_headers) . "\n# DELETE THIS LINE --- START DATA HERE AND USE APPROPRIATE DELIMITER/VALUE SEPARATOR KEY #";
```

Comment:
This looks much better :+1: But I would remove the word "KEY" since I think it might be confusing. I think "Use appropriate delimiter/value separator" is clear enough!

```suggestion
    $fileheaders = implode($file_delimiter[0], $column_headers) . "\n# DELETE THIS LINE --- START DATA HERE AND USE APPROPRIATE DELIMITER/VALUE SEPARATOR #";
```